provider/src/main/java/com/synopsys/integration/alert/provider/blackduck/web/project/ProjectFilterCustomFunctionAction.java

```diff
@@ -0,0 +1,112 @@
+import com.synopsys.integration.alert.common.persistence.model.ConfigurationModel;
+import com.synopsys.integration.alert.common.provider.state.StatefulProvider;
+import com.synopsys.integration.alert.common.rest.HttpServletContentWrapper;
+import com.synopsys.integration.alert.common.rest.model.FieldModel;
+import com.synopsys.integration.alert.common.security.authorization.AuthorizationManager;
+import com.synopsys.integration.alert.provider.blackduck.BlackDuckProperties;
+import com.synopsys.integration.alert.provider.blackduck.BlackDuckProvider;
+import com.synopsys.integration.blackduck.http.client.BlackDuckHttpClient;
+import com.synopsys.integration.blackduck.service.BlackDuckServicesFactory;
+import com.synopsys.integration.blackduck.service.dataservice.ProjectService;
+import com.synopsys.integration.log.Slf4jIntLogger;
+
+@Component
+public class ProjectFilterCustomFunctionAction extends CustomFunctionAction<BlackDuckProjectOptions> {
+    private static final String MISSING_PROVIDER_ERROR = "Provider name is required to retrieve projects.";
+    private static final ActionResponse<BlackDuckProjectOptions> NO_PROJECT_OPTIONS = new ActionResponse<>(HttpStatus.OK, new BlackDuckProjectOptions(List.of()));
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final ConfigurationAccessor configurationAccessor;
+    private final BlackDuckProvider blackDuckProvider;
+
+    @Autowired
+    public ProjectFilterCustomFunctionAction(AuthorizationManager authorizationManager, DescriptorMap descriptorMap, FieldValidationUtility fieldValidationUtility,
+        ConfigurationAccessor configurationAccessor, BlackDuckProvider blackDuckProvider) {
+        super(ProviderDistributionUIConfig.KEY_CONFIGURED_PROJECT, authorizationManager, descriptorMap, fieldValidationUtility);
+        this.configurationAccessor = configurationAccessor;
+        this.blackDuckProvider = blackDuckProvider;
+    }
+
+    @Override
+    public ActionResponse<BlackDuckProjectOptions> createActionResponse(FieldModel fieldModel, HttpServletContentWrapper servletContentWrapper) {
+        String providerName = fieldModel.getFieldValue(ChannelDistributionUIConfig.KEY_PROVIDER_NAME).orElse("");
+        if (StringUtils.isBlank(providerName)) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, MISSING_PROVIDER_ERROR);
+        }
+
+        return fieldModel.getFieldValue(ProviderDescriptor.KEY_PROVIDER_CONFIG_ID)
+                   .map(Long::parseLong)
+                   .map(this::getBlackDuckProjectsActionResponse)
+                   .orElse(NO_PROJECT_OPTIONS);
+    }
+
+    private ActionResponse<BlackDuckProjectOptions> getBlackDuckProjectsActionResponse(Long blackDuckGlobalConfigId) {
+        try {
+            Optional<ConfigurationModel> optionalBlackDuckGlobalConfig = configurationAccessor.getConfigurationById(blackDuckGlobalConfigId);
```

Comment:
The same functionality is in BlackDuckProviderDataAccessor, you might want to use that rather than implementing the same logic again here.